### PR TITLE
feat(core): allow passing `undefined` without needing to include it i…

### PR DESCRIPTION
**Pull request by JoostK**
…n the type argument of `input`

This commit introduces an overload for `input` to accept `undefined` as initial value if only options are needed to be provided, inferring an input of type `T|undefined`. Prior to this change, the type argument as specified needed to include `|undefined` explicitly even though that isn't necessary when passing options isn't needed.

**Note:** this is an experiment to determine if such change could impact existing code, it is yet to be determined if this is the direction we want to take.

Relates to #53909

### CODE REVIEW

File: packages/core/src/authoring/input/input.ts [hunk 1]
```diff
@@ -51,6 +51,8 @@ export interface InputFunction {
   <T>(): InputSignal<T | undefined>;
   /** Declares an input of type `T` with an explicit initial value. */
   <T>(initialValue: T, opts?: InputOptionsWithoutTransform<T>): InputSignal<T>;
+  /** Declares an input of type `T|undefined` without an initial value, but with input options */
+  <T>(initialValue: undefined, opts: InputOptionsWithoutTransform<T>): InputSignal<T | undefined>;
```
Comment (JoostK) [inline packages/core/src/authoring/input/input.ts:55]: I decided to make `opts` required if an explicit `undefined` is passed, making `input<string>(undefined)` a type error, with the intention that that should always be achieved using just `input<string>()`.

**Comment by ArielGueta** [comment]
The most difficult thing in the migration is if you have an input:

`@Input() foo: string;`

and you migrate it to signal input:

`foo = input<string | undefined>(undefined)`

The type is now "broken" because the consumers of this `input` supports only string and not expect to get `undefined`. We have many of these in our application.

**Comment by JoostK** [comment]
> The most difficult thing in the migration is if you have an input:
> 
> `@Input() foo: string;`
> 
> and you migrate it to signal input:
> 
> `foo = input<string | undefined>(undefined)`
> 
> The type is now "broken" because the consumers of this `input` supports only string and not expect to get `undefined`. We have many of these in our application and this is the reason we are not able to migrate currently.

That's unrelated to the changes here. The original input declaration is unsafe as the input may not be assigned, making it `undefined`. `input.required` avoids the introduction of `undefined`.

**Comment by ArielGueta** [comment]
We can't use the required because it's not required. I'm not saying that this is the correct way things should have been done but Angular was "forgiven" in that matter and it's now difficult to migrate. I'm sure there are more applications that did the same.

File: packages/core/src/authoring/input/input.ts (hunk 1 above)
Comment (devversion) [reply packages/core/src/authoring/input/input.ts:55]: I wonder if that is not confusing. TS error messages are not super great, but I like the incentive to encourage the shorthand.

**Review by devversion — APPROVED** [review]

**Comment by devversion** [comment]
> We can't use the required because it's not required. I'm not saying that this is the correct way things should have been done but Angular was "forgiven" in that matter and it's now difficult to migrate. I'm sure there are more applications that did the same.

I guess it depends on whether your application is built using `--strict` / `--strictPropertyInitialization`. In the example above, the input may also be `undefined`. So it's actually unsafe/brittle already, as @JoostK pointed out.

If you were to convert from `@Input` to `input()`, and would not fix the type to include `undefined`, or use a required input— then you'd effectively choose to do this: `input<string>(undefined as any)`

**Comment by devversion** [comment]
[TGP](https://fusion2.corp.google.com/presubmit/670143365/OCL:670143365:BASE:670110696:1725284173343:a06e452;groups=/targets) green

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

**Review by thePunderWoman — APPROVED** [review]
reviewed-for: public-api

**Pull request closed without merging**